### PR TITLE
[Java] Fix for issue #3638: Generated Java APIs incorrectly encode commas in certain query parameters

### PR DESCRIPTION
**Pull request by michaelkourlas**
This PR is a fix for #3638. Commas are now not percent-encoded when they are used as a delimiter in collection query parameters in the Java jersey1 and okhttp-gson APIs. A similar fix was made to the JavaScript API in #4725.

As part of this fix, a new API client method called `parameterToPair` has been added. This method converts query parameters to a list containing a single pair.

In addition, the behaviour of the API client method `parameterToPairs` has changed. The `value` parameter of this method must now always be a collection, and the `collectionFormat` parameter must now always specify a valid collection format. The values of the pairs returned are also now percent-encoded.

The Java API template now calls `parameterToPairs` when the collection format is defined for a query parameter, and `parameterToPair` if it is not. This should work because the DefaultCodegen class appears to ensure that templates receive a collection format for a query parameter ("csv" if none is explicitly specified in the spec) if and only if the parameter is of type "array" (or "object", apparently, but that type is not valid for parameters). Other templates (such as JavaScript) already seem to rely on this behaviour.

The `buildUrl` method has been updated to take into account these changes by placing collection query parameter values directly into the URL without percent encoding.

The Java unit tests for the `parameterToPairs` have been updated to take the new behaviour of the method into account. New tests have been added for the new `parameterToPair` method.

This PR was made against the 2.3.0 branch because it changed several method signatures, as in the case of PR #4725.

### CODE REVIEW

**Comment by wing328** [comment]
@michaelkourlas when you've time, would you please rebase on the latest 2.3.0 to resolve the merge conflicts?

**Comment by michaelkourlas** [comment]
@wing328 The build was failing on the Angular petstore client, which I never touched. When I tried the merge again it started failing on the Go petstore client.

**Comment by wing328** [comment]
@michaelkourlas thanks for the rebase. Don't worry about the CI failures (usually intermittent). I'll take a look

File: samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java [hunk 1]
```diff
@@ -203,25 +246,12 @@ public void testParameterToPairsWhenValueIsEmptyStrings() throws Exception {
...
-
-    @Test
     public void testParameterToPairsWhenValueIsCollection() throws Exception {
         Map<String, String> collectionFormatMap = new HashMap<String, String>();
         collectionFormatMap.put("csv", ",");
         collectionFormatMap.put("tsv", "\t");
         collectionFormatMap.put("ssv", " ");
-        collectionFormatMap.put("pipes", "\\|");
-        collectionFormatMap.put("", ","); // no format, must default to csv
-        collectionFormatMap.put("unknown", ","); // all other formats, must default to csv
```
Comment (wing328) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: @michaelkourlas is it correct to say that the default is no longer csv (",") when unknown or no format is provided?

Comment (michaelkourlas) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: @wing328 I think parameterToPairs still technically defaults to csv in those cases. If you want I can add the tests back.

However, I think this behaviour is wrong. I believe it is impossible to get that far without a defined collection format; if it's not defined DefaultCodegen appears to define it as "csv" for you. As such I don't think we need to handle the empty string case. Moreover, the invalid case should throw an error -- we shouldn't just silently assume csv. Don't we validate the spec before generating APIs from it first?

Comment (wing328) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: @michaelkourlas please put those test back when you've time.

I agree the behaviour is arguable but prefer to still keep the test cases before we change the behaviour.

> Don't we validate the spec before generating APIs from it first?

The current approach is to throw errors/warnings for invalid spec while trying to generate a functional clients. The goal is to provide a functional client that at least compiles right out of the box and the users can fix the spec or customize the setting (e.g. collection format) later.

Comment (michaelkourlas) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: Okay, I've gone ahead and added the tests back.

**Comment by michaelkourlas** [comment]
@wing328 Not sure why the AppVeyor build is failing; I didn't touch the C# stuff.

**Comment by wing328** [comment]
@michaelkourlas I'll take a look.

**Comment by michaelkourlas** [comment]
@wing328 Any idea when this will be merged?

**Comment by wing328** [comment]
@michaelkourlas sorry for the delay. Will take a look now...

**Pull request merged**